### PR TITLE
[autograd.Function] Fix torch.compile w/ once_differentiable leads to opaque graph break

### LINKED ISSUE #106893: [autograd.Function] torch.compile w/ once_differentiable leads to opaque graph break
### 🐛 Describe the bug

Some internal models use once_differentiable. The error message is very opaque:
```py
import torch
from typing import *
from torch.autograd.function import once_differentiable


class ScaleGradient(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    @once_differentiable
    def backward(ctx, grad):
        return grad * 0.5

x = torch.randn(3, requires_grad=True)
def f(x):
    return ScaleGradient.apply(x)
output = torch.compile(f, backend='eager', fullgraph=True)(x)
```
returns
```
Unsupported: inline in skipfiles: ScaleGradient.backward  | wrapper /raid/rzou/pt/debug-cpu2/torch/aut
ograd/function.py

from user code:
   File "<ipython-input-20-149cb1010897>", line 18, in f
    return ScaleGradient.apply(x)
  File "/raid/rzou/pt/debug-cpu2/torch/_dynamo/variables/misc.py", line 241, in trampoline_autograd_bw
d
    return fn_cls.backward(*args, **kwargs)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

```

cc @ezyang @gchanan @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov 

### Error logs

_No response_

### Minified repro

_No response_

### Versions

main

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #106893

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/misc.py [hunk 1]
```diff
@@ -456,6 +456,11 @@ def call_method(
             return self.call_apply(tx, args, kwargs)
         elif name == "backward":
             with tx.strict_translation_mode():
+                # backward function can be decorated by torch.autograd.function.once_differentiable,
+                # however, torch.compile doesn't support higher order gradients. And we can assume that
+                # the grad_mode during the backwards pass is False, so just inline the original function.
+                if hasattr(self.fn_cls.backward, "__wrapped__"):
+                    self.fn_cls.backward = self.fn_cls.backward.__wrapped__
```
Comment (zou3519) [inline torch/_dynamo/variables/misc.py:463]: once_differentiable might add `__wrapped__`, but it's not the only way to get `__wrapped__`. functools.wraps adds the `__wrapped__` attribute.

Is there some way to target this change to once_differentiable?

**Review by zou3519 — CHANGES_REQUESTED** [review]
I think this would allow other decorators other than once_differentiable on the backward() that we may not want allowed

File: torch/_dynamo/variables/misc.py (hunk 1 above)
Comment (yanboliang) [reply torch/_dynamo/variables/misc.py]: I added another check on top of this, to check if this function is wrapped by the function defined in ```torch.autograd.function```. Right now only ```once_differentiable``` is the case. Does it looks good to you?

Comment (yanboliang) [reply torch/_dynamo/variables/misc.py]: If users wrap ```backward``` function in their own file, the above check won't pass.

File: torch/_dynamo/variables/misc.py [hunk 2]
```diff
@@ -456,6 +457,15 @@ def call_method(
...
             with tx.strict_translation_mode():
+                # backward function can be decorated by torch.autograd.function.once_differentiable,
+                # however, torch.compile doesn't support higher order gradients. And we can assume that
+                # the grad_mode during the backwards pass is False, so just inline the original function.
+                if hasattr(
+                    self.fn_cls.backward, "__wrapped__"
+                ) and self.fn_cls.backward.__code__.co_filename == module_dir(
+                    torch.autograd.function
+                ):
+                    self.fn_cls.backward = self.fn_cls.backward.__wrapped__
```
Comment (zou3519) [inline torch/_dynamo/variables/misc.py:468]: This is better, but:
Why are we replacing self.fn_cls.backward with `self.fn_cls.backward.__wrapped__` ? We'd like to inline through once_differentiable wrapping the backward, not just the backward

Comment (yanboliang) [reply torch/_dynamo/variables/misc.py:468]: > once_differentiable has different behavior depending on if the gradients have requires_grad=True or not. In theory we don't actually know if the gradients have requires_grad=True or not at the time of the forward pass, but we may actually be able to assume they don't because the rest of torch.compile assumes that...

I was trying to inline the ```once_differentiable``` wrapped function, but we can't guarantee it passes L572 in Dynamo tracing. Then I found actually ```once_differentiable``` wrapped function just does L570 actually, which is the same as inlining the original backward function.
https://github.com/pytorch/pytorch/blob/9a9232956f90b0adf53049badf91a4309fb29a87/torch/autograd/function.py#L566-L573

Comment (yanboliang) [reply torch/_dynamo/variables/misc.py:468]: We can't access ```grad.requires_grad``` in Dynamo tracing.

Comment (zou3519) [reply torch/_dynamo/variables/misc.py:468]: The HOP tracing of the autograd.Function backward should set torch.is_grad_enabled to False, because PT2 doesn't support double backwards. It doesn't do this (because there is a bug).

Comment (yanboliang) [reply torch/_dynamo/variables/misc.py:468]: Ah, you are right! I think we should figure out why grad is enabled during Dynamo tracing backward graph.

Comment (zou3519) [reply torch/_dynamo/variables/misc.py:468]: You want to do something like the following https://github.com/pytorch/pytorch/pull/108686/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR971-R977 (this was my original once_differentiable PR that got reverted :P)

**Review by zou3519 — APPROVED** [review]
thank you!
